spdk exception handling

baum opened the pull request:
### spdk exception handling
The reason behind this change is to offer a dependable asynchronous notification of the termination of the spdk process. This will be achieved while collecting troubleshooting information and ensuring a graceful shutdown of the gateway process.
- pytests
  - using Python "with" GatewayServer
  - ini options
  - add spdk exception test
- extract _stop_spdk
- SIGCHLD signal handler
  - waitpid query of exit code
  - remove prctl code
  - terminate the process by raising SystemExit

### example
- [CI run](https://github.com/ceph/ceph-nvmeof/actions/runs/5819443618/job/15778322841):
```
------------------------------ Captured log setup ------------------------------
ERROR    control.server:server.py:220 Unable to initialize SPDK: 
 Timeout: file '/var/tmp/spdk.sock' not found after 60.0 seconds.
ERROR    control.server:server.py:165 SPDK return code: -4
ERROR    control.server:server.py:166 SPDK stdout:
 None
ERROR    control.server:server.py:167 SPDK stderr:
 None
```

### previous discussion

- [here](https://github.com/ceph/ceph-nvmeof/pull/176#discussion_r1293369972)

Commit:
spdk exception handling

- pytests
  - using Python "with" GatewayServer
  - ini options
  - add spdk exception test
- extract _stop_spdk
- SIGCHLD signal handler
  - waitpid query of exit code
  - remove prctl code
  - terminate the process by raising SystemExit

Co-authored-by: Ilya Dryomov <idryomov@gmail.com>
Signed-off-by: Alexander Indenbaum <aindenba@redhat.com>

baum commented:
> There is similar logic in `__exit__` and there a timeout is passed to `communicate`. Should it be passed here too to avoid hanging if SPDK is hung?
> 
> Or perhaps `_exception_spdk` could be melded in `__exit__` is that stdout/stderr are grabbed at all times instead of just in these two cases?

@idryomov 🖖

Before this change the output: (see [here](https://github.com/ceph/ceph-nvmeof/actions/runs/5855493074/job/15873839071?pr=176)):

```
ERROR    control.server:server.py:198 Unable to initialize SPDK: 
 Error while connecting to /var/tmp/spdk.sock
Is SPDK application running?
Error details: Invalid or non-existing address: '/var/tmp/spdk.sock'
``` 

Adding timeout to `communicate` to avoid potential indefinite hangs is a good idea 👍.
Thank you for pointing to `__exit__`, not sure though it is called since  the `GatewayServer` is not used with `with` statement.
WDYT?

idryomov commented:
> Thank you for pointing to `__exit__`, not sure though it is called since the `GatewayServer` is not used with `with` statement.

I see it being used with `with` in `control/__main__.py`:
https://github.com/ceph/ceph-nvmeof/blob/31a7239b28010f125a0175ba4f9186b142bd805e/control/__main__.py#L33-L36

idryomov commented on control/server.py:
```diff
@@ -147,6 +159,14 @@ def _add_server_listener(self):
+        timeout = self.config.getfloat("spdk", "timeout")
+        stdout, stderr = self.spdk_process.communicate(timeout=timeout)
+        return_code = self.spdk_process.returncode
+        self.logger.error(f"SPDK return code: {return_code}")
```
I would suggest "exit code" (Python's terminology is unfortunate here):
```suggestion
        self.logger.error(f"SPDK exit code: {return_code}")
```

trociny reviewed (commented):
Just as a note, we already tried to solve this issue in https://github.com/ceph/ceph-nvmeof/pull/81 just by increasing the default for conn_retries to 10, which is effectively about 2 sec timeout, because the  rpc client uses 0.2 sec sleeps between retries.

It looks like 10 retries (2 sec) is not enough for the tests?

I am actually fine with the current change, though not very happy with the approach, but I was not very happy with my "fix" too. 

I would like to mention though that it will not work for the cases when a previous spdk process crashed leaving the dead socket file. From this POV it would be better to check not just the file exist but that you may actually connect to it, but then you just will duplicate the rpc client code.

Ideally I would like the rpc client code was more flexible so we could just set a larger sleep interval instead of the hardcoded  0.2 sec, or if the sleep interval dynamically increased on retries.

epuertat commented on control/server.py:
```diff
@@ -41,6 +42,17 @@ def callable():
+        if os.path.exists(file_path):
+            return
+
+        if timeout is not None and time.time() - start_time >= timeout:
+            raise Exception(f"Timeout: file '{file_path}' not found after {timeout} seconds.")
```
```suggestion
        if timeout is not None and (elapsed_time := time.time() - start_time) >= timeout:
            raise Exception(f"Timeout: file '{file_path}' not found after {elapsed_time} seconds.")
```

baum replied:
@epuertat 🖖, not sure the suggested syntax works, since, In Python, the assignment statement itself does not return a value.  Python treats an assignment as a statement rather than an expression:
```python
Python 3.7.9 (v3.7.9:13c94747c7, Aug 15 2020, 01:31:08)
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import time
>>> start_time = time.time()
>>> if (elapsed := time.time() - start_time) >= 0:
  File "<stdin>", line 1
    if (elapsed := time.time() - start_time) >= 0:
                ^
SyntaxError: invalid syntax
>>> if (elapsed = time.time() - start_time) >= 0:
  File "<stdin>", line 1
    if (elapsed = time.time() - start_time) >= 0:
                ^
SyntaxError: invalid syntax
>>>
```

epuertat replied:
It works in Python >=3.8 (btw Python 3.7 EoL ended a couple of months ago).

epuertat commented on control/server.py:
```diff
@@ -41,6 +42,17 @@ def callable():
 
     return callable
 
+def wait_for_file(file_path, timeout=None, interval=1):
+    start_time = time.time()
+
+    while True:
+        if os.path.exists(file_path):
+            return
```
I don't remember if we'll have to enforce another condition, but this `wait_till_<something>()` methods are quite common, so we could just expose it as a generic checker:
```suggestion
def wait_for_condition(condition_method: Callable[[], bool], error_message=None, timeout=None, interval=1):
    start_time = time.time()

    while True:
        if condition_method:
            return
...

wait_for_condition(lambda: os.path.exists(spdk_file), ...)              
```

epuertat commented on control/server.py:
```diff
@@ -41,6 +42,17 @@ def callable():
+            return
+
+        if timeout is not None and time.time() - start_time >= timeout:
+            raise Exception(f"Timeout: file '{file_path}' not found after {timeout} seconds.")
```
I know this is an [OS-related Exception](https://docs.python.org/3/library/exceptions.html#TimeoutError), but I find that non-generic `Exception`s are always friendlier to debug:
```suggestion
            raise TimeoutError(f"After {elapsed_time} seconds: {error_message}")
```

epuertat commented on control/server.py:
```diff
@@ -147,6 +159,14 @@ def _add_server_listener(self):
+        stdout, stderr = self.spdk_process.communicate(timeout=timeout)
+        return_code = self.spdk_process.returncode
+        self.logger.error(f"SPDK return code: {return_code}")
+        self.logger.error(f"SPDK stdout:\n {stdout}")
+        self.logger.error(f"SPDK stderr:\n {stderr}")
```
```suggestion
        if stdout:
            self.logger.error(f"SPDK stdout:\n {stdout}")
        if stderr:
            self.logger.error(f"SPDK stderr:\n {stderr}")
```

epuertat replied:
BTW, maybe it's just me but I hate multi-line logs, as they complicate log regex/parsing :(

What I did in the past was either:
```python
self.logger.error(f"SPDK stderr:\n {stderr.replace('\n', '\\n')}")
```
Or
```python
for line in stderr.split():
    self.logger.error(f"SPDK stderr: {line}")
```

epuertat reviewed (commented):
Great improvement. This part was causing lots of headaches to users and QA engineers, as SPDK failures were masked by this.

epuertat commented:
> Adding timeout to `communicate` to avoid potential indefinite hangs is a good idea +1. Thank you for pointing to `__exit__`, not sure though it is called since the `GatewayServer` is not used with `with` statement. WDYT?

I think I already tried that and timeout didn't work as I initially expected. As this is being (pseudo)-daemonized,[^1] there's no way to distinguish a real timeout from the daemon simply waiting for pipe IO. If we put a `timeout` we'll be simply finishing the `nvmf_tgt` daemon after a given period.

[^1]: https://stackoverflow.com/a/473702 and https://pypi.org/project/python-daemon/

trociny commented on control/server.py:
```diff
@@ -147,6 +159,14 @@ def _add_server_listener(self):
+        return_code = self.spdk_process.returncode
+        self.logger.error(f"SPDK return code: {return_code}")
+        self.logger.error(f"SPDK stdout:\n {stdout}")
+        self.logger.error(f"SPDK stderr:\n {stderr}")
```
I believe currently you will always get `None` for stdout and stderr. To read output you would need to add `stderr=subprocess.PIPE, stdout=subprocess.PIPE` params to `subprocess.Popen` when starting the spdk. But we intentionally removed these in efb94c5ebfefd9a61926a778d8e2249039e43e1e, because the spdk may block when it produces a lot of output, which is never read from pipe, and reaches the pipe buffer limit. So currently the spdk just prints to stderr and stdout.

epuertat commented on control/server.py:
```diff
@@ -41,6 +42,17 @@ def callable():
 
     return callable
 
+def wait_for_condition(condition_method: Callable[[], bool], error_message=None, timeout=None, interval=1):
```
Now I see the type annotations... We should [enable mypy or any other static type checker](https://github.com/ceph/ceph-nvmeof/issues/188).

epuertat commented on control/server.py:
```diff
@@ -204,6 +214,38 @@ def _start_spdk(self):
+            try:
+                os.remove(rpc_socket)
+            except Exception as e:
+                self.logger.error(f"An error occurred while removing rpc socket {rpc_socket}: {e}")
```
Are we intentionally muting any exception here?

baum replied:
how?

epuertat replied:
If you catch an exception but don't re-raise it again, the error will just consist of an error message, but the program flow will continue without changes (I personally feel that the `except` clause should always force to specify a `pass` or `raise` built-int to make it explicit whether the muting is intentional or not).

baum replied:
Yeah, I could relate to this. Please consider, this code is potentially executed within exception handling code, such as in the case of  `TimeoutError`, always during application tear down. So it seems natural to handle exceptions, avoiding reraising in this case. Hope it makes sense.

epuertat reviewed (approved):
LGMT! Just a few comments!

baum commented:
@idryomov @epuertat @trociny Thank you for your help, ideas, and insights
- extracted _stop_spdk()
- changed pytests to use `with` syntax
- will squash the commits into one

WDYT?

baum commented:
@idryomov @epuertat @trociny 🖖
Reworked spdk exception handling introducing:
- SIGCHLD signal handler, remove prctl logic
- additional  pytests matrix  `spdk_exception` test

Your thoughts are welcome 😀

trociny commented on control/server.py:
```diff
@@ -30,17 +31,38 @@
+            return
 
+        if timeout is not None and time.time() - start_time >= timeout:
+            raise TimeoutError(f"Timeout: {error_message} after {timeout} seconds.")
```
nit: with `error_message=None` (which is the default) it will not look very nice.

epuertat replied:
I wouldn't repeat the `Timeout` part, since exception messages are always prefixed by exception names:
```py
raise TimeoutError("hello error")
TimeoutError: hello error
```
```suggestion
            raise TimeoutError(f"{error_message} (after {timeout} seconds)")
```
Example:
```py
>>> raise TimeoutError(f"{error_message} (after {timeout} seconds)")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TimeoutError: File not found (after 5 seconds)
```

baum replied:
Thank you Ernesto. Removed [wait_for_condition](https://github.com/ceph/ceph-nvmeof/pull/182#discussion_r1299620078) for now per Mykola comment

trociny commented on control/server.py:
```diff
@@ -72,15 +94,11 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, traceback):
         """Cleans up SPDK and server instances."""
+        if exc_type is not None:
+            self.logger.error("GatewayServer exception occurred:", exc_info=(exc_type, exc_value, traceback))
```
nit: (here and in all other cases) I would like we keep the 80 characters line limit, breaking longer lines where it is possible.

epuertat replied:
BTW, `logging` already provides a logger method `.exception()` which equals to `.error()` plus traceback dump.

trociny commented on control/server.py:
```diff
@@ -180,6 +200,8 @@ def _start_spdk(self):
             f" conn_retries: {conn_retries}, timeout: {timeout}",
         })
         try:
+            wait_for_condition(lambda: os.path.exists(spdk_rpc_socket), f"file '{spdk_rpc_socket}' not found", timeout)
```
As I said previously, I don't like very much we are duplicating the spdk client code here, while our implementation is less reliable (can not detect a dead socket) and potentially adds 1 sec delay on start (due to 1 sec check interval). I see you have added the socket file cleanup on termination, but it still may not work for all cases, like the host crash.

Actually I don't mind to have this, though personally I would prefer to rely on the spdk for this stuff, i.e. I would just increase the `conn_retries`. I think I would even remove `conn_retries` from the config, and just calculated it something like this (assuming the spdk client retries every 0.2 sec): `conn_retries = min(timeout / 0.2, 10)`.

trociny replied:
sorry, ofc I meant  `conn_retries = max(timeout / 0.2, 10)` here.

baum replied:
@trociny 🖖 agree, if JSON RPC client can handle the race condition where socket file does not exists yet, we could definitely remove wait_for_condition(),  just was not sure about that.

trociny commented on control/server.py:
```diff
@@ -204,6 +226,39 @@ def _start_spdk(self):
+        try:
+            timeout = self.config.getfloat("spdk", "timeout")
+            stdout, stderr = self.spdk_process.communicate(timeout=timeout)
+            for pair in [(stdout, "stdout"), (stderr, "stderr")]:
```
Also, as I said previously, I believe the stdout and stderr will always be None here, so why bother with this?

baum replied:
@trociny 🖖 I think you're right, however I feel uncomfortable treating .communicate() as returning void, ignoring stderr/stdout 😀. this code does not print a thing with current Popen argument, but might produce some useful output if codebase changes. WDYT?

trociny replied:
I don't see anything wrong with ignoring the communicate return value in this case if we expect it to be always None.

trociny commented on control/server.py:
```diff
@@ -204,6 +226,39 @@ def _start_spdk(self):
+        # Terminate spdk process
+        signal.signal(signal.SIGCHLD, signal.SIG_DFL)
+        if return_code is not None:
+            self.logger.error(f"SPDK({rpc_socket}) pid {self.spdk_process.pid} already terminated, exit code: {return_code}")
```
nit: why do we need `rpc_socket` in this log message (and below)? I would rather have it like `"SPDK({pid}) ..."`

baum replied:
@trociny 🖖 rpc socket is logged just for a narrow case of multi_gateway test, to differentiate between gatewayA and gatewayB. I would not mind promoting [gateway_name](https://github.com/baum/ceph-nvmeof/blob/21b77ce84ad6364d26a3f671b80189b13b220c51/control/server.py#L87) to be a self member and print it instead. WDYT?

trociny replied:
The gateway name looks more useful for me than the rpc socket, which I expect usually to be the same in production. Maybe we could add the gateway name as a log prefix?

trociny reviewed (commented):
@baum In general I like the idea of making handling the spdk and gateway termination more reliable. Thank you for working on this.

epuertat commented on .github/workflows/build-container.yml:
```diff
@@ -55,7 +55,7 @@ jobs:
       fail-fast: false
       matrix:
-          test: ["cli", "state", "multi_gateway"]
+          test: ["cli", "state", "multi_gateway", "spdk_exception"]
```
Nit: rather than after the bug name, I would name the test suite after:
* The python module (in this case `server`),
* The stage at which this happens (`start_up`)

That way you can add further related tests to that suite without having to rename it.

epuertat commented on control/server.py:
```diff
@@ -30,17 +31,38 @@
 
-libc = ctypes.CDLL(ctypes.util.find_library("c"))
-PR_SET_PDEATHSIG = 1
+def wait_for_condition(condition_method: Callable[[], bool], error_message=None, timeout=None, interval=1):
```
Given you have been so nice to type this signature, why not completely doing it?

```suggestion
def wait_for_condition(condition_method: Callable[[], bool], error_message: Optional[AnyStr]=None, timeout: Optional[float]=None, interval: float=1):
```

baum replied:
Thank you Ernesto. Removed [wait_for_condition](https://github.com/ceph/ceph-nvmeof/pull/182#discussion_r1299620078) for now per Mykola comment

epuertat commented on control/server.py:
```diff
@@ -163,9 +181,11 @@ def _start_spdk(self):
-            self.spdk_process = subprocess.Popen(cmd,
-                                                 preexec_fn=set_pdeathsig(
-                                                     signal.SIGTERM))
+            # install SIGCHLD handler
+            signal.signal(signal.SIGCHLD, sigchld_handler)
+
+            # start spdk process
+            self.spdk_process = subprocess.Popen(cmd)
```
Did you consider using an [existing daemon implementation](https://peps.python.org/pep-3143/#reference-implementation)?

baum replied:
Thank you for the idea, @epuertat , I will 👀 into it.

trociny commented:
@baum You have removed your "wait for socket file" code but haven't increased the connection retries. I think we need one of these things. As I understand you observed that the current number of retries was not enough?

baum commented:
> @baum You have removed your "wait for socket file" code but haven't increased the connection retries. I think we need one of these things. As I understand you observed that the current number of retries was not enough?

@trociny 🖖 reviewed the spdk client code, thank you for pointers. We could derive reconnect retries from timeout configuration value then  😀.

epuertat commented on control/server.py:
```diff
@@ -204,6 +207,37 @@ def _start_spdk(self):
+        assert self.spdk_process is not None # should be verified by the caller
+
+        return_code = self.spdk_process.returncode
+        rpc_socket = self.config.get("spdk", "rpc_socket")
```
One thing that I'm worried about using `configparser` is that most parsing errors will occur lazily, at the time the value is accessed. I know that `rpc_socket` is already fetched during start-up phase, but I'd personally enforce required params & defaults at initialization. I just [opened this issue](https://github.com/ceph/ceph-nvmeof/issues/191) to tackle that.

baum replied:
Guess we're safe here. In general, digital sheep dream of Rust style safeness. I recommend beginning by establishing the schema for the configuration class, along with mandatory default values. Agree on validation of the configuration file during creation or initial conf file read to ensure a 'fail-fast' approach. While the INI format and the ('section', 'key') concept are commonly used, considering a more versatile and generic class might provide greater flexibility.

trociny commented on control/server.py:
```diff
@@ -163,18 +164,20 @@ def _start_spdk(self):
         log_level = self.config.get("spdk", "log_level")
-        conn_retries = self.config.getint_with_default("spdk",
-                                                       "conn_retries", 10)
+        # connect timeout: spdk client reties 5 times per sec
```
s/reties/retries/

trociny commented on control/server.py:
```diff
@@ -30,17 +29,22 @@
+    exit_code = os.waitstatus_to_exitcode(wait_status)
 
+    # GW process should exit now
+    raise SystemExit(f"spdk subprocess terminated pid {pid=} {exit_code=}")
```
"pid" word is not needed here, because we have with it "spdk subprocess terminated pid pid=..."

trociny reviewed (commented):
In general the current version LGTM, though I have not tested it very well to see how well it works for different cases.

Currently, my only question is don't we produce too much backtraces and how useful are they? Here is an example of my log when it failed to start the spdk:

```
[2023-08-22 12:47:39.629184] init.c: 602:spdk_env_init: *ERROR*: Failed to initialize DPDK
[2023-08-22 12:47:39.629208] app.c: 351:app_setup_env: *ERROR*: Unable to initialize SPDK env
ERROR:control.server:GatewayServer: GSIGCHLD received signum=17 stack=
  File "/usr/lib64/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,

  File "/usr/lib64/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)

  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/__main__.py", line 35, in <module>
    gateway.serve()

  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/server.py", line 97, in serve
    self._start_spdk()

  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/server.py", line 186, in _start_spdk
    self.spdk_rpc_client = rpc_client.JSONRPCClient(

  File "/home/mgolub/ceph/ceph-nvmeof.upstream/spdk/python/spdk/rpc/client.py", line 64, in __init__
    time.sleep(0.2)

ERROR:control.server:GatewayServer exception occurred:
Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/spdk/python/spdk/rpc/client.py", line 94, in _connect
    raise socket.error("Invalid or non-existing address: '%s'" % addr)
OSError: Invalid or non-existing address: '/var/tmp/spdk.sock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/spdk/python/spdk/rpc/client.py", line 60, in __init__
    self._connect(addr, port)
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/spdk/python/spdk/rpc/client.py", line 96, in _connect
    raise JSONRPCException("Error while connecting to %s\n"
spdk.rpc.client.JSONRPCException: Error while connecting to /var/tmp/spdk.sock
Is SPDK application running?
Error details: Invalid or non-existing address: '/var/tmp/spdk.sock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/__main__.py", line 35, in <module>
    gateway.serve()
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/server.py", line 97, in serve
    self._start_spdk()
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/server.py", line 186, in _start_spdk
    self.spdk_rpc_client = rpc_client.JSONRPCClient(
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/spdk/python/spdk/rpc/client.py", line 64, in __init__
    time.sleep(0.2)
  File "/home/mgolub/ceph/ceph-nvmeof.upstream/control/server.py", line 47, in sigchld_handler
    raise SystemExit(f"spdk subprocess terminated pid {pid=} {exit_code=}")
SystemExit: spdk subprocess terminated pid pid=5771 exit_code=1
INFO:control.server:Terminating SPDK(adonis) pid 5771...
INFO:control.server:Exiting the gateway process.
spdk subprocess terminated pid pid=5771 exit_code=1
```

trociny commented on control/server.py:
```diff
@@ -163,18 +164,20 @@ def _start_spdk(self):
+        conn_retries = int(timeout * 5)
         self.logger.info({
             f"Attempting to initialize SPDK: rpc_socket: {spdk_rpc_socket},",
             f" conn_retries: {conn_retries}, timeout: {timeout}",
```
BTW, this message looks like below:
```
INFO:control.server:{'Attempting to initialize SPDK: rpc_socket: /var/tmp/spdk.sock,', ' conn_retries: 300, timeout: 60.0'}
```
I think the curly brackets around the formatted message string is wrong here.

baum commented:
> Currently, my only question is don't we produce too much backtraces and how useful are they? Here is an example of my log when it failed to start the spdk:

Agree, removed the stack trace logging in sigchld handler. Your insights and feedback have been incredibly valuable in improving the code and making it better.  Thank you, Mykola 🖖

trociny commented on control/server.py:
```diff
@@ -62,25 +65,22 @@ def __init__(self, config):
+        if not self.name:
+            self.name = socket.gethostname()
+        self.logger.info(f"Starting gateway {self.name}")
+        self.logger.name
```
@baum is it correct?

baum replied:
thank you!

trociny reviewed (approved):
Otherwise LGTM (though not tested).
